### PR TITLE
Rework our CI specifications

### PR DESCRIPTION
**Pull request by thanodnl**
This doesn't change anything functionally for Citus or our CI. Instead it is a quality of life change to make it easier to maintain CI and how we build.

In general this patch accomplishes the following
 - move build and test scripts into the Citus tree instead of our CI tree.
 - deduplicate convoluted jobs definition by a couple generic build and test runs
   - standardized the output
   - standardized coredumps
   - made all coredumps and regression diffs available for download

A big part of this refactor is in the `the-process` repository. https://github.com/citusdata/the-process/pull/48

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -5,47 +5,32 @@ orbs:
 
 jobs:
 
-  build-11:
+  build:
+    description: Build the citus extension
+    parameters:
+      postgres:
```
Comment (SaitTalhaNisanci) [inline .circleci/config.yml:11]: Minor: to remove the mental mapping, we can call this parameter `postgres_major` or `pg_major`

File: .circleci/config.yml [hunk 2]
```diff
@@ -5,47 +5,32 @@ orbs:
...
+        description: postgres major version building citus for
+        type: integer
+      image:
+        description: docker image to use for the build
+        type: string
+        default: citus/extbuilder
+      tag:
```
Comment (SaitTalhaNisanci) [inline .circleci/config.yml:18]: Minor: maybe `image_tag`

File: .circleci/config.yml [hunk 3]
```diff
@@ -5,47 +5,32 @@ orbs:
...
-  build-13:
-    docker:
-      - image: 'citus/extbuilder:13.0'
-    steps:
-      - checkout
-      - run:
-          name: 'Configure, Build, and Install'
-          command: build-ext
+          command: |
+            ./ci/build-extension.sh
```
Comment (SaitTalhaNisanci) [inline .circleci/config.yml:28]: maybe build-citus.sh?

File: .circleci/config.yml [hunk 4]
```diff
@@ -99,459 +85,228 @@ jobs:
...
-      - codecov/upload:
-          flags: 'test_11,vanilla'
 
-  test-11_check-mx:
-    docker:
-      - image: 'citus/exttester:11.9'
+  test-pg-upgrade:
+    description: Runs postgres upgrade tests
+    parameters:
+      from:
```
Comment (SaitTalhaNisanci) [inline .circleci/config.yml:92]: `old_pg_major`

File: .circleci/config.yml [hunk 5]
```diff
@@ -99,459 +85,228 @@ jobs:
...
+        description: 'postgres major version to use before the upgrade'
+        type: integer
+      to:
+        description: 'postgres major version to upgrade to'
+        type: integer
+      image:
+        description: 'docker image to use as for the tests'
+        type: string
```
Comment (SaitTalhaNisanci) [inline .circleci/config.yml:100]: `new_pg_major`

File: .circleci/config.yml [hunk 6]
```diff
@@ -99,459 +85,228 @@ jobs:
...
-           install-and-test-ext --target check-citus-upgrade --citus-pre-tar /install-pg11-citusv8.1.0.tar
-           install-and-test-ext --target check-citus-upgrade --citus-pre-tar /install-pg11-citusv8.2.0.tar
-           install-and-test-ext --target check-citus-upgrade --citus-pre-tar /install-pg11-citusv8.3.0.tar
-
-           install-and-test-ext --target check-citus-upgrade-mixed --citus-pre-tar /install-pg11-citusv8.0.0.tar
-           install-and-test-ext --target check-citus-upgrade-mixed --citus-pre-tar /install-pg11-citusv8.1.0.tar
-           install-and-test-ext --target check-citus-upgrade-mixed --citus-pre-tar /install-pg11-citusv8.2.0.tar
-           install-and-test-ext --target check-citus-upgrade-mixed --citus-pre-tar /install-pg11-citusv8.3.0.tar
+            # run make check-citus-upgrade for all citus versions
+            for citus_version in ${CITUS_VERSIONS}; do \
```
Comment (SaitTalhaNisanci) [inline .circleci/config.yml:196]: this comes from the container right? If so maybe we can have a comment here

**Review by SaitTalhaNisanci — COMMENTED** [review]
I guess we will need to update the required tests: `check-mx` -> `check-multi-mx` on github. `check-multi-mx` seems more consistent with the target name. Previously the job name was `check-mx` but the target was `check-multi-mx`.

File: .circleci/config.yml [hunk 7]
```diff
@@ -603,78 +359,196 @@ workflows:
...
-      - test-11_check-multi:
+      - test-citus:
+          name: 'test-11_check-multi'
+          postgres: 11
+          tag: 11.10-dev202012151914
+          make: check-multi
           requires: [build-11]
-      - test-11_check-vanilla:
+      - test-citus:
+          name: 'test-11_check-multi-mx'
```
Comment (thanodnl) [inline .circleci/config.yml:385]: revert back to old name of test. that will prevent us from having to change github settings

```suggestion
          name: 'test-11_check-mx'
```

File: .circleci/config.yml [hunk 8]
```diff
@@ -603,78 +359,196 @@ workflows:
...
-      - test-12_check-multi:
+      - test-citus:
+          name: 'test-12_check-multi'
+          postgres: 12
+          tag: 12.5-dev202012151914
+          make: check-multi
           requires: [build-12]
-      - test-12_check-vanilla:
+      - test-citus:
+          name: 'test-12_check-multi-mx'
```
Comment (thanodnl) [inline .circleci/config.yml:429]: revert back to old name of test. that will prevent us from having to change github settings

```suggestion
          name: 'test-12_check-mx'
```

File: .circleci/config.yml [hunk 9]
```diff
@@ -603,78 +359,196 @@ workflows:
...
-      - test-13_check-multi:
+      - test-citus:
+          name: 'test-13_check-multi'
+          postgres: 13
+          tag: 13.1-dev202012151914
+          make: check-multi
           requires: [build-13]
-      - test-13_check-vanilla:
+      - test-citus:
+          name: 'test-13_check-multi-mx'
```
Comment (thanodnl) [inline .circleci/config.yml:485]: revert back to old name of test. that will prevent us from having to change github settings

```suggestion
          name: 'test-13_check-mx'
```

File: .circleci/config.yml [hunk 10]
```diff
@@ -193,6 +193,7 @@ jobs:
           name: 'Install and test citus upgrade'
           command: |
             # run make check-citus-upgrade for all citus versions
+            # the image has ${CITUS_VERSIONS} set with all verions it contains the binaries of
```
Comment (SaitTalhaNisanci) [inline .circleci/config.yml:196]: versions

File: .circleci/config.yml [hunk 11]
```diff
@@ -203,6 +204,7 @@ jobs:
             done;
 
             # run make check-citus-upgrade-mixed for all citus versions
+            # the image has ${CITUS_VERSIONS} set with all verions it contains the binaries of
```
Comment (SaitTalhaNisanci) [inline .circleci/config.yml:207]: versions

**Review by SaitTalhaNisanci — APPROVED** [review]

**Pull request merged**